presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpGenericPattern.java

```diff
@@ -0,0 +1,435 @@
+            int offset = m.search(0, source.length(), Option.DEFAULT);
+            return offset != -1;
+        }
+        return false;
+    }
+
+    private void re2jCheckReplacementString(String replacementString, int groupCount)
+    {
+        //checking some special cases in the replacement string to make it compatible to joni processing
+        //Currently only handles exception cases similar to joni processing
+        int idx = 0;
+        int replacementStringLength = replacementString.length();
+        while (idx < replacementStringLength) {
+            char nextChar = replacementString.charAt(idx);
+            if (nextChar == '$') {
+                idx++;
+                if (idx == replacementStringLength) {
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacementString);
+                }
+                nextChar = replacementString.charAt(idx);
+                if (nextChar != '{') {
+                    int backref = nextChar - '0';
+                    if (backref < 0 || backref > 9) { // not using checkArgument because `.toStringUtf8` is expensive
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacementString);
+                    }
+                    if (groupCount < backref) {
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: unknown group " + backref);
+                    }
+                }
+            }
+            else {
+                if (nextChar == '\\') {
+                    idx++;
+                    if (idx == replacementString.length()) {
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacementString);
+                    }
+                }
+            }
+            idx++;
+        }
+    }
+
+    public Slice replace(Slice source, Slice replacement)
+    {
+        if (selectedRegexLibrary == RegexLibrary.RE2J) {
+            String inputString =  source.toStringUtf8();
+            String replacementString = replacement.toStringUtf8();
+            com.google.re2j.Matcher matcher = re2jPattern.matcher(inputString);
+
+            re2jCheckReplacementString(replacementString, re2jPattern.groupCount());
+            String replacedString;
+
+            try {
+                replacedString = matcher.replaceAll(replacementString);
+            }
+            catch (Exception exp) {
+               //any exception here is assumed to be due to INVALID_FUNCTION_ARGUMENT
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacementString);
+            }
+
+            SliceOutput sliceOutput = new DynamicSliceOutput(source.length() + replacement.length() * 5);
+            sliceOutput.appendBytes(replacedString.getBytes(), 0, replacedString.length());
```

Comment:
Method to be used is: `Slices.utf8Slice`. Using `replacedString.getBytes()` is not correct, because it encodes bytes using platform's default charset, which is not necessarily UTF8. Good practice is to always specify encoding explicitly.